pkg/agent/openflow/pod_connectivity.go

```diff
@@ -155,16 +161,92 @@ func (f *featurePodConnectivity) initFlows() []binding.Flow {
 		// Pod IP will take care of routing the traffic to destination Pod.
 		flows = append(flows, f.l3FwdFlowToLocalPodCIDR()...)
 	}
+	if f.enableTrafficControl {
+		flows = append(flows, f.trafficControlCommonFlows()...)
+	}
 	return flows
 }
 
 func (f *featurePodConnectivity) replayFlows() []binding.Flow {
 	var flows []binding.Flow
 
 	// Get cached flows.
-	for _, cachedFlows := range []*flowCategoryCache{f.nodeCachedFlows, f.podCachedFlows} {
+	for _, cachedFlows := range []*flowCategoryCache{f.nodeCachedFlows, f.podCachedFlows, f.tcCachedFlows} {
 		flows = append(flows, getCachedFlows(cachedFlows)...)
 	}
 
 	return flows
 }
+
+// trafficControlMarkFlows generates the flows to mark the packets that need to be redirected or mirrored.
+func (f *featurePodConnectivity) trafficControlMarkFlows(sourceOFPorts []uint32, targetOFPort uint32, direction v1alpha2.Direction, action v1alpha2.TrafficControlAction) []binding.Flow {
+	cookieID := f.cookieAllocator.Request(f.category).Raw()
+	var actionRegMark *binding.RegMark
+	if action == v1alpha2.ActionRedirect {
+		actionRegMark = TrafficControlRedirectRegMark
+	} else if action == v1alpha2.ActionMirror {
+		actionRegMark = TrafficControlMirrorRegMark
+	}
+	var flows []binding.Flow
+	for _, port := range sourceOFPorts {
+		if direction == v1alpha2.DirectionIngress || direction == v1alpha2.DirectionBoth {
+			// This generates the flow to mark the packets destined for a provided port.
+			flows = append(flows, TrafficControlTable.ofTable.BuildFlow(priorityNormal).
+				Cookie(cookieID).
+				MatchRegFieldWithValue(TargetOFPortField, port).
+				Action().LoadToRegField(TrafficControlTargetOFPortField, targetOFPort).
+				Action().LoadRegMark(actionRegMark).
+				Action().NextTable().
+				Done())
+		}
+		// This generates the flow to mark the packets sourced from a provided port.
+		if direction == v1alpha2.DirectionEgress || direction == v1alpha2.DirectionBoth {
+			flows = append(flows, TrafficControlTable.ofTable.BuildFlow(priorityNormal).
+				Cookie(cookieID).
+				MatchInPort(port).
+				Action().LoadToRegField(TrafficControlTargetOFPortField, targetOFPort).
+				Action().LoadRegMark(actionRegMark).
+				Action().NextTable().
+				Done())
+		}
+	}
+	return flows
+}
+
+// trafficControlReturnClassifierFlow generates the flow to mark the packets from traffic control return port.
+func (f *featurePodConnectivity) trafficControlReturnClassifierFlow(returnOFPort uint32) binding.Flow {
+	return ClassifierTable.ofTable.BuildFlow(priorityNormal).
+		Cookie(f.cookieAllocator.Request(f.category).Raw()).
+		MatchInPort(returnOFPort).
+		Action().LoadRegMark(FromTCReturnRegMark).
+		Action().GotoStage(stageRouting).
```

Comment:
Why not go to stageSwitching since the dst MAC in the packet is set before output to the redirected port?

Reply:
I'll update the comments here

```
// trafficControlReturnClassifierFlow generates the flow to mark the packets from traffic control return port and forward
// the packets to stageRouting directly. Note that, for the packets which are originally to be output to a tunnel port,
// value of NXM_NX_TUN_IPV4_DST for the returned packets needs to be loaded in stageRouting.
```